src/sources/file/line_agg.rs

```diff
@@ -0,0 +1,519 @@
+    config: Config,
+
+    /// Line per filename.
+    buffers: HashMap<Filename, BytesMut>,
+
+    /// Draining queue. We switch to draining mode when we get `None` from
+    /// the inner stream. In this mode we stop polling `inner` for new lines
+    /// and just flush all the buffered data.
+    draining: Option<Vec<(Bytes, Filename)>>,
+
+    /// A queue of filename timeouts.
+    timeouts: DelayQueue<Filename>,
+
+    /// A queue of filenames with expired timeouts.
+    expired: VecDeque<Filename>,
+}
+
+impl<T> LineAgg<T> {
+    pub(super) fn new(inner: T, config: Config) -> Self {
+        Self {
+            inner,
+
+            config,
+
+            draining: None,
+            buffers: HashMap::new(),
+            timeouts: DelayQueue::new(),
+            expired: VecDeque::new(),
+        }
+    }
+}
+
+impl<T> Stream for LineAgg<T>
+where
+    T: Stream<Item = (Bytes, Filename), Error = ()>,
+{
+    /// `Bytes` - the message data; `Filename` - file name.
+    type Item = (Bytes, Filename);
+    type Error = ();
+
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        loop {
+            // If we're in draining mode, short circut here.
+            if let Some(to_drain) = &mut self.draining {
+                if let Some((line, src)) = to_drain.pop() {
+                    return Ok(Async::Ready(Some((line, src))));
+                } else {
+                    return Ok(Async::Ready(None));
+                }
+            }
+
+            // Check for keys that have hit their timeout.
+            while let Ok(Async::Ready(Some(expired_key))) = self.timeouts.poll() {
+                self.expired.push_back(expired_key.into_inner());
+            }
+
+            match self.inner.poll() {
+                Ok(Async::Ready(Some((line, src)))) => {
+                    // Handle the incoming line we got from `inner`. If the
+                    // handler gave us something - return it, otherwise continue
+                    // with the flow.
+                    if let Some(val) = self.handle_line(line, src) {
+                        return Ok(Async::Ready(Some(val)));
+                    }
+                }
+                Ok(Async::Ready(None)) => {
+                    // We got `None`, this means the `inner` stream has ended.
+                    // Start flushing all existing data, stop polling `inner`.
+                    self.draining =
+                        Some(self.buffers.drain().map(|(k, v)| (v.into(), k)).collect());
+                }
+                Ok(Async::NotReady) => {
+                    // We didn't get any lines from `inner`, so we just give
+                    // a line from the expired lines queue.
+                    if let Some(key) = self.expired.pop_front() {
+                        if let Some(buffered) = self.buffers.remove(&key) {
+                            return Ok(Async::Ready(Some((buffered.freeze(), key))));
+                        }
+                    }
```

Comment:
I don't understand why this condition wouldn't be checked unconditionally before polling the inner. If I'm understanding right, this would result in newer lines being emitted before expired lines, which will have been originally read in earlier. I see you copied this logic from the original implementation, so I'm not really asking about your change, just trying to understand the reasoning here.

Reply:
Honestly, I didn’t look into this part much, cause it’s a part of the older implementation. I decided not to change anything unrelated to the extended multi-line support.
I personally also found this odd while I was getting to know the code, and I agree that this should be in reverse order. Otherwise we don’t provide proper backpressure and the expired lines may never even flush at all if the load is super-high.

Reply:
I'd avoid changing that in this iteration. We better create a separate issue. There should be many more tests for the timeout and expiration logic, and once those are done - we can safely alter details like this.

Reply:
Agreed

Reply:
Created #1870.